### PR TITLE
[DRAFT] Final load gas particles

### PR DESCRIPTION
**Pull request by anschaible**
This branch implements also the loading of gas particles. It is based upon changing requirements in the branch load-gas-particles. It should at some point merged to the main, but there are still two things to be done:

- In rubix/core/data.py the data are loaded. There we should have also the option to only load a subset of the particles (https://github.com/ufuk-cakir/rubix/blob/c7ac30ce7ba96ffb1d22a8779990e17f140feae3/rubix/core/data.py#L241). However, if I try this the execution of the pipeline, the resuling spectra of the data cube are all 0 (see noteboot spectra_calculation.ipynb). At the moment I do not see, where this error arrises and how to fix it, the pipeline does not throw an error. I would be happy, if someone else could have a look into this part of the code.

- The unit tests are not changed for the gas loading, so currently some of the test fails. If you agree with this idea to load the gas particle in this way, I will change the unittests.

### CODE REVIEW

File: rubix/core/telescope.py [hunk 1]
```diff
@@ -44,13 +44,17 @@ def get_spaxel_assignment(config: dict) -> Callable:
...
 
-    def spaxel_assignment(input_data: dict) -> dict:
-        pixel_assignment = square_spaxel_assignment(
-            input_data["coords"], spatial_bin_edges
-        )
-        input_data["pixel_assignment"] = pixel_assignment
-        input_data["spatial_bin_edges"] = spatial_bin_edges
-        return input_data
+    def spaxel_assignment(rubixdata: object) -> object:
+        for particle_name in ["stars", "gas"]:
```
[the remark this answers is not in the document]
Comment (anschaible) [reply rubix/core/telescope.py:49]: vmap is not able to handle arrays/list of strings and throws also errors, if I use the index of the stars gas list. I now created a helper function that is then called twice. This avoids an forloop, however, it is not a real parallelization. I am not sure, if this is satisfying enough?

File: rubix/core/data.py [hunk 2]
```diff
@@ -10,6 +12,149 @@
...
+#
+#    def get_attributes(self) -> list:
+#        """
+#        Returns the filtered attributes.
+#        """
+#        return self.attributes
+
+
+# Registering the dataclass with JAX for automatic tree traversal
+@partial(jax.tree_util.register_pytree_node_class)
```
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/core/data.py:42]: can we make this an issue and do it in the future maybe? Then we can finally merge. @ufuk-cakir

File: rubix/core/telescope.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TobiBu) [reply rubix/core/telescope.py]: maybe we leave the for loop so we can loop over other particle types like BHs etc as well.

**Review by ufuk-cakir — APPROVED** [review]
look good

**Comment by anschaible** [comment]
Okay, then this brach is finally in a stage to be merged into the main

**Pull request merged**